packages/lesswrong/lib/sql/Query.ts

```diff
@@ -0,0 +1,515 @@
+import Table from "./Table";
+import { Type, JsonType, ArrayType } from "./Type";
+
+/**
+ * Arg is a wrapper to mark a particular value as being an argument for the
+ * query. When compiled, the value will be placed into the `args` array, and
+ * a `$n` reference will be placed in the appropriate place in the SQL string.
+ */
+class Arg {
+  public typehint = "";
+
+  constructor(public value: any, type?: Type) {
+    // JSON arrays make node-postgres fall over, but we can work around it
+    // with a special-case typehint
+    if (Array.isArray(value) && value[0] && typeof value[0] === "object") {
+      if (type instanceof JsonType) {
+        this.value = JSON.stringify(this.value);
+        this.typehint = "::JSONB";
+      } else {
+        this.typehint = "::JSONB[]";
+      }
+    }
+  }
+}
+
+/**
+ * The Atom represents one 'part' of a Postgres query. This could be a literal
+ * string of SQL, an argument, a sub-query, or a table (which will be compiled
+ * to its name).
+ */
+export type Atom<T extends DbObject> = string | Arg | Query<T> | Table;
```

Comment:
A lot of the places where I get confused about what's going on are to do with Atoms sometimes being objects and sometimes being strings. I can see how this makes things easier to do (once you get down to low level stuff you can just build strings which is easier to think about) but I think it would be good to refactor over time to have less plain string building and more "syntax tree" type compiling to avoid bugs like leading/trailing commas, missing parens etc